### DISCUSSION TITLE
Recipe Revision History - Track changes over time

### DISCUSSION THREAD

**Discussion by jakecovert** [mealie-recipes/mealie#4584] [Ideas]
I would be really nice to go back and see how a recipe has changed over time.  A la git revisions...

**Comment by ChipSkylark37** [comment]
This would be interesting. 

Sort of like a timeline of changes and why you may have deviated from the original recipe. Especially considering taking a recipe from any source and tweaking it to be your own. 

Example Timeline

On mm/dd/yyyy:
- changed ingredient (4) eggs to (5) eggs
- changed step 1 from "old value" to "new value"
- added comment "added another egg for xx reason"

**Comment by jakecovert** [comment]
That would be really nice.  As it stands now, I'm taking a recipe, and adding instances of creation to that timeline with notes.

But the NEXT time I go and try and cook that recipe, there's no easy way to reference what I did last time.  

This could be as simple as a pop-up / badge count of comments, so historical tweaks can be referenced.  

But I love revision history, cause I'm a nerd.  

On considerations, these are two asks really:

a) git core recipe changes over time: would LOVE this.
b) Some way to refer to the notes that are captures when timeline entries are made.  "Hey, this recipe was made 3 times. See the notes here..."

**Comment by ChipSkylark37** [comment]
I tend to just adjust the single recipe that's in mealie but you raised an interesting point. 

Most of the time I am making adjustments, its to tweak it for all future plans, though if it's just a one-time change (whoops, added 4 eggs instead of 5), then I may just put it the timeline using "I made this" or the comment section of the recipe itself. 

I know during the survey not a lot of people are using timeline, and with the inability to edit the timeline and that it's not really easily referenced, another way to audit & track comments / revisions would be welcomed.

**Comment by parumpum** [comment]
I like the idea of tracking changes, but I think first we'd have to resolve how updates happen. I've noticed that if you modify ingredients in a recipe you don't actually need to click Save.

If we can fix that, it would be easy to capture the before/after. 

Recipes do have a reference to their timeline events. I think it would be good to display those either with comments section or somewhere, wouldn't be too difficult to set that up.

**Comment by ChipSkylark37** [comment]
Hmm that's an interesting find. It doesn't happen in all cases though.

One note is I use pushover to notify if any recipe changes, so I'll list that in my tests and outcomes. 

Entering edit mode on a recipe and changing one of my ingredients from 100g to 101g:

1. Click **X** when in "editing"

Expected: 
- no changes made
- no notification

Actual:
- **Results in saving the new value**
- No notification of recipe update

2. click **save** on "editing"

Expected: 
- Results in saving the new value
- Receive notification of recipe being updated

Actual:
- Results in saving the new value
- Receive notification of recipe being updated

3. Click off page intentionally without clicking **X** or **Save** and clicking **cancel** to discard changes

Expected: 
- No changes made
- No notification

Actual:
- No changes made
- No notification

4. Click off page intentionally without clicking **X** or **Save** and clicking **OK** to save changes

Expected: 
- Changes made
- Notification

Actual:
- Changes made
- Notification

So it really seems like **scenario 1** is the issue and clicking the X button doesn't "cancel" the changes for ingredients, but most other methods work properly. 

5. Not clicking anything, other than modifying ingredient and then making misc other changes that typically result in a save (changing star rating). You actually see the loading bar for a second as it saves that change (stars) but if you then exit the page, it does prompt for "unsaved changes". 
- This is as expected

----------------------

Other comments:

An "audit trail" makes sense to have that lists:
- Old Value
- New Value
- Date of change
- Time of change
- User that made change

Comments would be nice, but not expected, I'd rather just put that in the comment section myself. Unless you had something like "Comment Type" that you could drop down to select things like "Recipe Update, Comment on Taste, Misc Notes" but that's going much farther than a simple audit trail / tracking changes. 

Should clicking the **X** after making changes and deciding not to save them result in the same "warning" of unsaved changes? Or should it just cancel the changes without a notification? My thought is a warning, since it's right next to the save button and I've definitely accidentally clicked it before. 

Truth be told, I personally find the "cancel" and "ok" confusing as well. I'd expect if I tried to leave the page and I didn't save my changes it would 1. Prompt me like it does but 2. Clicking **ok** would keep me on the page and not just "save". I have to stop and read the dialog each time because to make sure I know what I'm doing.

**Comment by parumpum** [comment]
Testing out the sort of git revision style system for recipes by creating an immutable RecipeHistory table that will include

-  Recipe ID
-  User ID
-  Timestamp
-  Reference to the "parent" revision to track the chain of changes
-  Changed fields/values
 
If this is the first revision stored, the changed fields/values would contain the whole recipe. Future revisions could just be a JSON dict that contains the changed items to save storage space over time. May need to put this into a file as a whole recipe may end up too much data for a single database column.

Going to focus on just the backend portions to start with. Will build the table, associated ORM, and tests.

**Reply by ChipSkylark37** [reply]
The way I have seen this be implemented in SQL databases is something like:

**Recipe (Table)**
RecipeRecID, (other values relating to recipe, ingredients, steps, etc,) RecipeAuditTrailRecID

**RecipeAuditTrail (Table)** (Likely the new thing that will need to be created to store the changes)
RecipeAuditTrailRecID
RecipeRecID (Reference to the "parent" revision to track the chain of changes)
User ID
Timestamp
Old Value
New Value

So you have your Two Primary Keys that relate back to each other and can be used to reference.

There can be multiple rows in the Recipe Table that have different RecipeAuditTrailRecID, along with the timestamp that they occured, and then it does that lookup to the 'RecipeAuditTrail' table to pull in what those changes were.

**Reply by parumpum** [reply]
I think our approaches are similar, mine does also include a relationship to the Recipes table at the ORM level. 

Storing the New Value alongside the Old Value would probably give a performance bump, not having to work as hard to calculate a diff when viewing revision history. Storage would increase, but we're just talking kilobytes here.

**Comment by DeepReef11** [comment]
Is anyone actively working on recipe revision history? 

I see @parumpum's feat/recipe-change-tracking branch from Jan 2025,  is that still being pursued?

I'd be interested in contributing to get this feature merged if the approach is acceptable to maintainers. @hay-kot would this kind of git-style revision  tracking be welcome as a PR?

**Comment by DeepReef11** [comment]
I have added this PR #7313 

It works fine for me. 

It has been rejected by a contributor with fair enough reason: it is generated by AI and I only did elaboration and testing. I have not done a deep dive on the actual code base. That being said, the feature is simple: save json copies of the different recipe version.

Missing features are: opt-in button and number of saved version to keep (would always keep original plus x most recent).
